### DISCUSSION TITLE
survey-html-form not capturing textarea values

### DISCUSSION THREAD

**Discussion by justinHCI** [jspsych/jsPsych#1095] [Q&A] [answered]
I am using a survey-html-form with several input types including numbers, selects, and textareas. When I save the values captured by the form however, the textareas are not appearing. 

I have a custom save function getting form names and values like this:
`var demographics = jsPsych.data.get().filter({trial_type: 'survey-html-form'}).values();`

but it does not have form name/value pairs for the textareas. A textarea in my survey looks like this:
`<div class='form' id = "highTime" name="stratLowHow">
        <label for='straHighTime'> For urgent trials, how did you decide when to deliver messages your partner? </label>
		<textarea id="stratHighTime" name="stratLowHow" cols=100 rows=5 form = "highTime" required></textarea></div>`

while a (working) number field looks like this: 
`<div class='form'>
        <label for='age'>Age: </label>
        <input type='number' id='age' name='age' min='18' max='100' required>
    </div>`

Any suggestions on how to get my textareas included when I save my data would be much appreciated. I'm a communications student working on my PhD so even this sort of coding is pretty tough for me to grasp.

**Comment by becky-gilbert** [comment]
Hi @justinHCI! I think the problem here is that you're assigning a form to the `<textarea>` element. The survey-html-form plugin will create its own form and then put all of your HTML inside of that, so when the trial is over, it looks for all of the values inside its own form. So I _think_ that's why your textarea data isn't being saved. 

I tested this by removing `form="HighTime"` from the `<textarea>` tag and the response is saving for me. Following the [textarea docs](https://www.w3schools.com/tags/tag_textarea.asp), I also added quotes around the `cols` and `rows` values, but I'm not sure whether that was causing any problems.
```html
<textarea id="stratHighTime" name="stratLowHow" cols="100" rows="5" required></textarea>
```

**Comment by justinHCI** [comment]
Thanks for the advice. Unfortunately this did not save my responses from textareas. To further troubleshoot the issue, I ran my experiment and on the screen after the survey in question I executed the following code in the browser console:
`jsPsych.data.get().filter({trial_type: 'survey-html-form'}).values();`

which returned an object containing a response time for the survey as well as responses to all the other forms in the survey, I suspect I'm doing things wrong in general by putting my selects and my number inputs within form elements, but those are all captured fine. I tried removing the form element that my textareas were in (while still specifying a name and id for those textareas) but that didn't help either.

I further tried specifying aform attribute for the textarea that matches the form id that the plugin creates ( form="jspsych-survey-html-form") in on of my text areas, but this also did not help.

I really appreciate your help on this.

**Comment by justinHCI** [comment, marked as answer]
So I did some more investigating and I found that, for some reason, my textareas lost their name attribute when the plugin displays the survey. I couldn't figure out how to make them keep their name attributes, so I made a custom plugin that is identitical to jspsych-survey-html-form but when serializing the array it checks for textareas and assigns them a name attribute from a hardcoded list. Hacky but gets the job done for me. Thanks again for taking a look at this for me!

**Reply by becky-gilbert** [reply]
Thanks for the update, and glad you figured the problem! That's very weird. It could be a bug in the plugin but hard to know because I can't replicate it. I can try again if you want to share a minimal example and let me know what browser you're using. And in any case I'm glad you figured out a work around.

**Reply by justinHCI** [reply]
Sure, I'd be happy to help the community by trying to replicate this bug. I tested this in both Edge and in Firefox and got the same result. I have an experiment.html file that welcomes participants and executes my script.js file. The script file has a bunch of things in its timeline, including a survey-html-form that screens participants to make sure they qualify for the study. It then has some trials involving videos, then near the end has my demographics and post-trial questionnaire. That questionnaire is defined like this:

````
var demographic = { 
    type: 'survey-html-form',
    preamble: "Thank you for taking part in the experiment today. We would now like you to complete the following questionnaire.",
    html: demo
````

and the demo.html file looks basically like this :

````
<div class='demographics'>
    <h1>Post-trial Questionnaire</h1>
    <div class='form'>
        <label for='age'>Age: </label>
        <input type='number' id='age' name='age' min='18' max='100' required>
    </div>
	<br><br>

<div class='form' >
        <label for='stratLowTime'> For non-urgent trials, how did you decide when to deliver messages your partner? </label>
	
	<textarea id="stratLowTime" name= "stratLowTimeName" cols="100" rows="5" required></textarea></div>
		  <br><br>
</div>
````

After that survey, the next timeline item is an html sheet debriefing the participant which has the following on_load value:
```
on_load: function () {
		prevent_leave = false;
		saveInfo('completion_code', ID);

		var results = jsPsych.data.get();
		var interaction = jsPsych.data.getInteractionData();
		var data = results.join(interaction);
		var name = ID 
		saveData(name, data.csv());
		
		   
        var demographics = jsPsych.data.get().filter({trial_type: 'survey-html-form'}).values();
        var merged = $.extend({},JSON.parse(demographics[0].responses),JSON.parse(demographics[1].responses));
        var csv = "";
        var headers = "";
        var values = "";
        let keysAmount = Object.keys(merged).length;
        let keysCounter = 0;
        for (var key in merged){
            headers += key + (keysCounter+1 < keysAmount ? ',' : '\r\n' );
            keysCounter++;
        }
        keysCounter = 0       
        for (var key in merged){
            values += "\"" + merged[key] + "\"" + (keysCounter+1 < keysAmount ? ',' : '\r\n' );
            keysCounter++;
        }
        csv = headers + values;
        var name = ID + "_demographics";
        saveData(name,csv);
        var valueList = [ID].concat(Object.values(merged));
        appendDemographics(valueList);
	}
};

```
saveInfo saves a participant's completion code to their local storage and associates it with their ID - their turkerID in the case that I'm running this on Amazon Mechanical Turk.

saveData is a function that does the following, calling a PHP file that POSTs the csv of data to my server.

```
function saveData(name, data) {
    var xhr = new XMLHttpRequest();
    xhr.open('POST', 'write_data.php'); 
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.send(JSON.stringify({
        filename: name,
        filedata: data
    }));
}
```

And appendDemographics is a function that does the following, calling a PHP file that POSTs a csv of just demographic data (not interaction data from the experiment) to another spot on the server.

```
function appendDemographics(values){
    var xhr = new XMLHttpRequest();
    xhr.open('POST', 'append_data.php'); // 'write_data.php' is the path to the php file described above.
    xhr.setRequestHeader('Content-Type', 'application/json');
    xhr.send(JSON.stringify({
        filedata: values
    }));
}
```

The outcome of all of this is that I get a main data csv (which includes both interaction data and survey data) and a survey only csv, each of which captures all of the responses in the screening questions (all of which are select inputs) and all of the data in the demo.html survey that isn't in a textarea. 

As I mentioned, I've confirmed that when demo.html survey is shown to the participant, in both Firefox and Edge, the elements that are number or select inputs keep their name attribute while the textarea elements do not keep their name and that is why the plugin was not saving the values entered in those elements.

Thanks much for taking a look at my code, I know how inelegant and hacky it is, so I'm especially grateful you're taking the time to comb through with me.

**Reply by becky-gilbert** [reply]
I'm not sure if this is the problem, but I noticed that you have a mix of single and double quotes for attribute values in your demo HTML. So when I wrapped this HTML in quotes to pass it to the survey-html-form `html` parameter, I had to change the `<textarea>` quotes from double to single. This works for me in Chrome, Firefox, and Edge (i.e. I can see my textarea response in the data):
```javascript
var trial = {
  type: 'survey-html-form',
  questions: [{prompt: '<span style="font-size:30px; position: absolute; top: 30%; with: 100%; left: 8%; text-align: center; ;">How much pollen do you think will be present on average in this garden? </span>', required: true, rows: 1, columns: 3}],
  html: "<div class='demographics'><h1>Post-trial Questionnaire</h1><div class='form'>"+
    "<label for='age'>Age: </label><input type='number' id='age' name='age' min='18' max='100' required>"+
    "</div><br><br>"+
    "<div class='form' >"+
    "<label for='stratLowTime'> For non-urgent trials, how did you decide when to deliver messages your partner? </label>"+
    "<textarea id='stratLowTime' name='stratLowTimeName' cols='100' rows='5' required></textarea></div>"+
    "<br><br></div>"
};
jsPsych.init({
  timeline: [trial],
  on_finish: function () {
    jsPsych.data.displayData();
  }
});
```

**Comment by justinHCI** [comment]
Great, yeah not sure if that's all it was but if so, that's an easy enough fix. I was passing demo.html to the survey-html-form as an address (I just had html: demo.html in the trial variable) so maybe my bad quotes get messed up when it's passed that way instead of directly. Hopefully that helps someone in the future!
